File: src/main/java/net/insomniakitten/glazed/block/GlassBlock.java
package net.insomniakitten.glazed.block;

/*
 *  Copyright 2018 InsomniaKitten
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *   you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 *   Unless required by applicable law or agreed to in writing, software
 *   distributed under the License is distributed on an "AS IS" BASIS,
 *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *   See the License for the specific language governing permissions and
 *   limitations under the License.
 */

import net.insomniakitten.glazed.Glazed;
import net.insomniakitten.glazed.GlazedVariant;
import net.minecraft.block.BlockGlass;
import net.minecraft.block.SoundType;
import net.minecraft.block.material.Material;
import net.minecraft.block.state.BlockStateContainer;
import net.minecraft.block.state.IBlockState;
import net.minecraft.client.util.ITooltipFlag;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.BlockRenderLayer;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.NonNullList;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.RayTraceResult;
import net.minecraft.world.Explosion;
import net.minecraft.world.IBlockAccess;
import net.minecraft.world.World;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

import javax.annotation.Nullable;
import java.util.List;

import static net.insomniakitten.glazed.GlazedVariant.PROPERTY;
import static net.insomniakitten.glazed.GlazedVariant.VARIANTS;

public final class GlassBlock extends BlockGlass {
    public GlassBlock() {
        super(Material.GLASS, true);
        setUnlocalizedName(Glazed.ID + ".glass_block");
        setSoundType(SoundType.GLASS);
        setCreativeTab(Glazed.TAB);
    }

    @Override
    @Deprecated
    public IBlockState getStateFromMeta(int meta) {
        return getDefaultState().withProperty(PROPERTY, VARIANTS[meta]);
    }

    @Override
    public int getMetaFromState(IBlockState state) {
        return state.getValue(PROPERTY).ordinal();
    }

    @Override
    @Deprecated
    public float getBlockHardness(IBlockState state, World world, BlockPos pos) {
        return state.getValue(PROPERTY).getHardness();
    }

    @Override
    public boolean onBlockActivated(World world, BlockPos pos, IBlockState state, EntityPlayer player, EnumHand hand, EnumFacing side, float hitX, float hitY, float hitZ) {
        return state.getValue(PROPERTY).onPlayerInteract(world, pos, player, side);
    }

    @Override
    public void onEntityWalk(World world, BlockPos pos, Entity entity) {
        world.getBlockState(pos).getValue(PROPERTY).onEntityWalk(world, pos, entity);
    }

    @Override
    public void onEntityCollidedWithBlock(World world, BlockPos pos, IBlockState state, Entity entity) {
        state.getValue(PROPERTY).onEntityCollide(world, pos, entity);
    }

    @Override
    public void getSubBlocks(CreativeTabs tab, NonNullList<ItemStack> items) {
        for (GlazedVariant variant : VARIANTS) {
            items.add(new ItemStack(this, 1, variant.ordinal()));
        }
    }

    @Override
    protected BlockStateContainer createBlockState() {
        return new BlockStateContainer(this, PROPERTY);
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void addInformation(ItemStack stack, World world, List<String> tooltip, ITooltipFlag flag) {
        tooltip.add(GlazedVariant.getDescription(stack));
    }

    @Override
    public int getLightValue(IBlockState state, IBlockAccess access, BlockPos pos) {
        return state.getValue(PROPERTY).getLightLevel();
    }

    @Override
    public boolean doesSideBlockRendering(IBlockState state, IBlockAccess world, BlockPos pos, EnumFacing side) {
        final IBlockState other = world.getBlockState(pos.offset(side));
        return other.getBlock() == this && other.getValue(PROPERTY) == state.getValue(PROPERTY);
    }

    @Override
    public float getExplosionResistance(World world, BlockPos pos, @Nullable Entity exploder, Explosion explosion) {
        return world.getBlockState(pos).getValue(PROPERTY).getResistance();
    }

    @Override
    public ItemStack getPickBlock(IBlockState state, RayTraceResult target, World world, BlockPos pos, EntityPlayer player) {
        return new ItemStack(this, 1, state.getValue(PROPERTY).ordinal());
    }

    @Override
    public boolean canRenderInLayer(IBlockState state, BlockRenderLayer layer) {
        return state.getValue(PROPERTY).getRenderLayer() == layer;
    }

    @Override
    public SoundType getSoundType(IBlockState state, World world, BlockPos pos, @Nullable Entity entity) {
        return state.getValue(PROPERTY).getSoundType();
    }

    @Override
    @SideOnly(Side.CLIENT)
    public boolean shouldSideBeRendered(IBlockState state, IBlockAccess world, BlockPos pos, EnumFacing side) {
        return !doesSideBlockRendering(state, world, pos, side);
    }
}

File: src/main/kotlin/net/insomniakitten/glazed/block/GlassBlock.kt
package net.insomniakitten.glazed.block

import net.insomniakitten.glazed.Glazed
import net.insomniakitten.glazed.GlazedVariant
import net.insomniakitten.glazed.GlazedVariant.VARIANTS
import net.insomniakitten.glazed.extensions.description
import net.insomniakitten.glazed.extensions.doesSideBlockRendering
import net.insomniakitten.glazed.extensions.plus
import net.insomniakitten.glazed.extensions.variant
import net.minecraft.block.BlockGlass
import net.minecraft.block.material.Material
import net.minecraft.block.state.BlockStateContainer
import net.minecraft.block.state.IBlockState
import net.minecraft.client.util.ITooltipFlag
import net.minecraft.creativetab.CreativeTabs
import net.minecraft.entity.Entity
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.util.BlockRenderLayer
import net.minecraft.util.EnumFacing
import net.minecraft.util.NonNullList
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.RayTraceResult
import net.minecraft.world.Explosion
import net.minecraft.world.IBlockAccess
import net.minecraft.world.World
import net.minecraftforge.fml.relauncher.Side
import net.minecraftforge.fml.relauncher.SideOnly

class GlassBlock : BlockGlass(Material.GLASS, true) {
    override fun getUnlocalizedName() = "block.${Glazed.ID}.glass_block"

    override fun getCreativeTabToDisplayOn() = Glazed.TAB

    override fun createBlockState() = BlockStateContainer(this, GlazedVariant)

    override fun getStateFromMeta(meta: Int) = defaultState + GlazedVariant(meta)

    override fun getMetaFromState(state: IBlockState) = state.variant.ordinal

    override fun getSoundType(state: IBlockState, world: World, pos: BlockPos, entity: Entity?) =
            state.variant.soundType

    override fun getBlockHardness(state: IBlockState, world: World, pos: BlockPos) =
            state.variant.hardness

    override fun getExplosionResistance(world: World, pos: BlockPos, entity: Entity?, explosion: Explosion) =
            world.getBlockState(pos).variant.resistance

    override fun getLightValue(state: IBlockState, access: IBlockAccess, pos: BlockPos) =
            state.variant.lightLevel

    override fun canRenderInLayer(state: IBlockState, layer: BlockRenderLayer) =
            state.variant.renderLayer == layer

    override fun canProvidePower(state: IBlockState) = state.variant.redstoneLevel > 0

    override fun getWeakPower(state: IBlockState, access: IBlockAccess, pos: BlockPos, side: EnumFacing) =
            state.variant.redstoneLevel

    override fun getSubBlocks(tab: CreativeTabs, items: NonNullList<ItemStack>) =
            VARIANTS.forEach { items += ItemStack(this, 1, it.ordinal) }

    @SideOnly(Side.CLIENT)
    override fun addInformation(stack: ItemStack, world: World?, tooltip: MutableList<String>, flag: ITooltipFlag) {
        stack.description.run(tooltip::add)
    }

    override fun doesSideBlockRendering(state: IBlockState, world: IBlockAccess, pos: BlockPos, side: EnumFacing) =
            world.getBlockState(pos.offset(side)).let { it.block == this && it.variant == state.variant }

    override fun getPickBlock(state: IBlockState, hit: RayTraceResult, world: World, pos: BlockPos, player: EntityPlayer) =
            ItemStack(this, 1, state.variant.ordinal)

    @SideOnly(Side.CLIENT)
    override fun shouldSideBeRendered(state: IBlockState, access: IBlockAccess, pos: BlockPos, side: EnumFacing) =
            !access.doesSideBlockRendering(pos.offset(side), side.opposite)
}
